src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultWorkspaceChangedRefresh.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using OmniSharp.Extensions.LanguageServer.Protocol.Server;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    /// <summary>
+    /// Sends a 'workspace\semanticTokens\refresh' request each time the project changes.
+    /// </summary>
+    internal class DefaultWorkspaceSemanticTokensRefreshTrigger : ProjectSnapshotChangeTrigger
+    {
+        private IWorkspaceSemanticTokensRefreshPublisher? _workspaceChangedPublisher;
+        private ProjectSnapshotManagerBase? _projectSnapshotManager;
+        private readonly IClientLanguageServer _clientLanguageServer;
+
+        internal DefaultWorkspaceSemanticTokensRefreshTrigger(IClientLanguageServer clientLanguageServer!!)
+        {
+            _clientLanguageServer = clientLanguageServer;
+        }
+
+        private ProjectSnapshotManagerBase ProjectSnapshotManager
+        {
+            get
+            {
+                if (_projectSnapshotManager is null)
+                {
+                    throw new InvalidOperationException("ProjectSnapshotManager accessed before Initialized was called.");
+                }
+
+                return _projectSnapshotManager;
+            }
+        }
+
+        private IWorkspaceSemanticTokensRefreshPublisher WorkspaceChangedPublisher
+        {
+            get
+            {
+                if (_workspaceChangedPublisher is null)
+                {
+                    throw new InvalidOperationException($"{WorkspaceChangedPublisher} accessed before Initialized was called.");
+                }
+
+                return _workspaceChangedPublisher;
+            }
+        }
+
+        public override void Initialize(ProjectSnapshotManagerBase projectManager)
+        {
+            _projectSnapshotManager = projectManager;
+            _workspaceChangedPublisher = GetWorkspaceSemanticTokensRefreshPublisher(_projectSnapshotManager);
+
+            ProjectSnapshotManager.Changed += ProjectSnapshotManager_Changed;
+        }
+
+        internal virtual IWorkspaceSemanticTokensRefreshPublisher GetWorkspaceSemanticTokensRefreshPublisher(ProjectSnapshotManagerBase projectManager)
+        {
+            var errorReporter = projectManager.Workspace.Services.GetRequiredService<ErrorReporter>();
+            return new DefaultWorkspaceSemanticTokensRefreshPublisher(_clientLanguageServer, errorReporter);
+        }
```

Comment:
Could register the publisher as a service and have all thise taken care of for you

Reply:
Might be worth registering an error reporter time in the language servers service collection too. We've been trying to get away from the roslyn services